File: ch7/polygon.ml
open Printf

(* From points_lines.ml *)

let eps = 1e-9
let near x y = abs_float (x -. y) < eps
let pi = Float.pi

let deg_to_rad d = d *. pi /. 180.
let rad_to_deg r = r *. 180. /. pi

type point = { x: float; y: float }
let point (x, y) = { x; y }

let dist a b = hypot (a.x -. b.x) (a.y -. b.y)

type vector = Vector of float * float

let to_vec a b = Vector(b.x -. a.x, b.y -. a.y)
let scale (Vector(a, b)) s = Vector(a *. s, b *. s)
let translate {x; y} (Vector(a, b)) = point(x +. a, y +. b)
let dot (Vector(a1, b1)) (Vector(a2, b2)) = a1 *. a2 +. b1 *. b2
let norm v = sqrt (dot v v)
let cross (Vector(a1, b1)) (Vector(a2, b2)) = a1 *. b2 -. a2 *. b1

let angle a o b =
  let oa = to_vec o a in
  let ob = to_vec o b in
  acos (dot oa ob /. (norm oa *. norm ob))

let collinear p q r = near (cross (to_vec p q) (to_vec p r)) 0.

(* End from points_lines.ml *)

let ccw p q r =
  let s = cross (to_vec p q) (to_vec p r) in
  if near s 0. then 0 else
  if s > 0. then 1 else -1

module Polygon : sig
  type t = point array
  val of_points         : point list -> t
  val points            : t -> point list
  val perimeter         : t -> float
  val area              : t -> float
  val area_alternative  : t -> float
  val is_convex         : t -> bool
  val in_polygon        : t -> point -> bool
  val cut_polygon       : t -> point -> point -> t
  val convex_hull       : point list -> t
end = struct
  (** Counter-clockwise order, the last point is the same as the first point *)
  type t = point array

  let of_points ps =
    let n = List.length ps in
    let arr = Array.make (n + 1) (point(0., 0.)) in
    ps |> List.iteri (fun i v -> arr.(i) <- v);
    arr.(n) <- arr.(0);
    arr
  
  let points p = p |> Array.to_list |> List.tl

  let perimeter p =
    let n = Array.length p in
    let result = ref 0. in
    for i = 0 to n - 2 do
      let a, b = p.(i), p.(i + 1) in
      result := !result +. dist a b
    done;
    !result

  let area p =
    let n = Array.length p in
    let result = ref 0. in
    for i = 0 to n - 2 do
      let a, b = p.(i), p.(i + 1) in
      result := !result +. (a.x *. b.y -. b.x *. a.y)
    done;
    abs_float !result /. 2.0

  let area_alternative p =
    let n = Array.length p in
    let o = point(0., 0.) in
    let result = ref 0. in
    for i = 0 to n - 2 do
      let a, b = p.(i), p.(i + 1) in
      result := !result +. cross (to_vec o a) (to_vec o b)
    done;
    abs_float !result /. 2.0

  let is_convex p =
    let n = Array.length p in
    if n <= 3 then false else begin
      let l, r = ref false, ref false in
      for i = 0 to n - 2 do
        let a, b, c = p.(i), p.(i + 1), p.(if i + 2 = n then 1 else i + 2) in
        let s = ccw a b c in
        if s = 1 then l := true else
        if s = -1 then r := true
      done;
      not (!l && !r)
    end
  
  let in_polygon p point =
    let n = Array.length p in
    if n <= 3 then false else begin
      let sum = ref 0. in
      let border = ref false in
      for i = 0 to n - 2 do
        let a, b = p.(i), p.(i + 1) in
        let side = ccw point a b in
        if side = 1 then
          sum := !sum +. angle a point b
        else if side = -1 then sum := !sum -. angle a point b
        else border := true
      done;
      !border || abs_float !sum > pi
    end

  let line_intersect_segment p q r s =
    let a = s.y -. r.y in
    let b = r.x -. s.x in
    let c = s.x *. r.y -. r.x *. s.y in
    let u = abs_float (a *. p.x +. b *. p.y +. c) in
    let v = abs_float (a *. q.x +. b *. q.y +. c) in
    point((p.x *. v +. q.x *. u) /. (u+.v), (p.y *. v +. q.y *. u) /. (u+.v))

  let cut_polygon p a b =
    let p' = ref [] in
    let n = Array.length p in
    for i = 0 to n - 1 do
      let left1 = cross (to_vec a b) (to_vec a p.(i)) in
      let left2 =
        if i <> n - 1 then cross (to_vec a b) (to_vec a p.(i + 1))
        else 0. in
      if left1 > -. eps then p' := p.(i) :: !p'; (* Q[i] is on the left of ab *)
      if left1 *. left2 < -.eps then   (* edge (Q[i], Q[i+1]) crosses line ab *)
        p' := line_intersect_segment p.(i) p.(i + 1) a b :: !p';
    done;
    if List.length !p' > 0 then begin
      let last = !p' |> List.fold_left (fun _ x -> x) (point(0., 0.)) in
      if last <> (List.hd !p') then p' := last :: !p'
    end;
    !p' |> List.rev |> Array.of_list

  (** Andrew's algorithm *)
  let convex_hull points =
    let points = points |> Array.of_list in
    points |> Array.sort (fun a b ->
      compare (a.x, a.y) (b.x, b.y));
    let n = Array.length points in
    let hull = Array.make (2 * n) (point(0., 0.)) in
    let k = ref 0 in
    for i = 0 to n - 1 do
      while !k >= 2 && ccw hull.(!k-2) hull.(!k-1) points.(i) <= 0 do
        decr k
      done;
      hull.(!k) <- points.(i);
      incr k
    done;
    let t = !k + 1 in
    for i = n - 2 downto 0 do
      while !k >= t && ccw hull.(!k-2) hull.(!k-1) points.(i) <= 0 do
        decr k
      done;
      hull.(!k) <- points.(i);
      incr k
    done;
    let result = Array.make (!k + 1) hull.(0) in
    for i = 0 to !k - 1 do result.(i) <- hull.(i) done;
    result
end

let () =
  let p = [(1, 1); (3, 3); (9, 1); (12, 4); (9, 7); (1, 7)]
          |> List.map (fun (x, y) -> point(float_of_int x, float_of_int y))
          |> Polygon.of_points in

  printf "Perimeter of polygon = %.2f\n" (Polygon.perimeter p);
  printf "Area of polygon = %.2f\n" (Polygon.area p);
  printf "Area of polygon = %.2f\n" (Polygon.area_alternative p);
  printf "Is convex = %b\n" (Polygon.is_convex p);

  (* the positions of P6 and P7 w.r.t the polygon
  7 P5--------------P4
  6 |                  \
  5 |                    \
  4 |   P7                P3
  3 |   P1___            /
  2 | / P6    \ ___    /
  1 P0              P2
  0 1 2 3 4 5 6 7 8 9 101112 *)

  let p6 = point(3., 2.) in
  printf "Point P6 is inside this polygon = %b\n"
    (Polygon.in_polygon p p6);
  let p7 = point(3., 4.) in
  printf "Point P7 is inside this polygon = %b\n"
    (Polygon.in_polygon p p7);
  
  (* cutting the original polygon based on line P[2] -> P[4] (get the left side)
  7 P5--------------P4
  6 |               |  \
  5 |               |    \
  4 |               |     P3
  3 |   P1___       |    /
  2 | /       \ ___ |  /
  1 P0              P2
  0 1 2 3 4 5 6 7 8 9 101112
  new polygon (notice the index are different now):
  7 P4--------------P3
  6 |               |
  5 |               |
  4 |               |
  3 |   P1___       |
  2 | /       \ ___ |
  1 P0              P2
  0 1 2 3 4 5 6 7 8 9 *)

  let p = Polygon.cut_polygon p (point(9., 1.)) (point(9., 7.)) in
  printf "Perimeter of polygon = %.2f\n" (Polygon.perimeter p);
  printf "Area of polygon = %.2f\n" (Polygon.area p);

  (* running convex hull of the resulting polygon (index changes again)
  7 P3--------------P2
  6 |               |
  5 |               |
  4 |   P7          |
  3 |               |
  2 |               |
  1 P0--------------P1
  0 1 2 3 4 5 6 7 8 9 *)

  let p = Polygon.convex_hull (Polygon.points p) in
  printf "Perimeter of polygon = %.2f\n" (Polygon.perimeter p);
  printf "Area of polygon = %.2f\n" (Polygon.area p);
  printf "Is convex = %b\n" (Polygon.is_convex p);
  printf "Point P6 is inside this polygon = %b\n" (Polygon.in_polygon p p6);
  printf "Point P7 is inside this polygon = %b\n" (Polygon.in_polygon p p7)

File: ch7/polygon.py
# we will convert the C++ version to Python soon



import math

EPS = 1e-9
# const double EPS = 1e-9;

# double DEG_to_RAD(double d) { return d*M_PI / 180.0; }

# double RAD_to_DEG(double r) { return r*180.0 / M_PI; }

class point:
  def __init__(self, x, y):                       # constructor
    self.x = x
    self.y = y
  
  def __lt__(self, b): return (self.x, self.y) < (b.x, b.y)
  
  def __str__(self): return "{} {}".format(self.x, self.y)
  def __hash__(self):return hash((self.x,self.y))
#   bool operator == (point other) const {
#    return (fabs(x-other.x) < EPS && (fabs(y-other.y) < EPS)); } 
#   bool operator <(const point &p) const {
#    return x < p.x || (abs(x-p.x) < EPS && y < p.y); } };

class vec:
  def __init__(self, x=0, y=0):
    self.x = x
    self.y = y

def toVec(a, b):
  return vec(b.x-a.x, b.y-a.y)
# struct vec { double x, y;  // name: `vec' is different from STL vector
#   vec(double _x, double _y) : x(_x), y(_y) {} };

# vec toVec(point a, point b) {       // convert 2 points to vector a->b
#   return vec(b.x-a.x, b.y-a.y); }

def dist(p1, p2):                                 # Euclidean distance
  return math.hypot(p1.x-p2.x, p1.y-p2.y)

# returns the perimeter of polygon P, which is the sum of
# Euclidian distances of consecutive line segments (polygon edges)
# def perimeter(P) return math.fsum(dist(P[i], P[i+1]) for i in range(len(P)-1)) 
def perimeter(P):
  ans = 0.0
  for i in range(len(P)-1):                       # note: P[n-1] = P[0]
    ans += dist(P[i], P[i+1])                     # as we duplicate P[0]
  return ans

def area(P):
  return math.fsum(cross(P[i], P[i+1]) for i in range(len(P)-1))/2
# // returns the area of polygon P
# double area(const vector<point> &P) {
#   double ans = 0.0;
#   for (int i = 0; i < (int)P.size()-1; ++i)      // Shoelace formula
#     ans += (P[i].x*P[i+1].y - P[i+1].x*P[i].y);
#   return fabs(ans)/2.0;                          // only do / 2.0 here
# }



def dot(a, b): return a.x * b.x + a.y * b.y
# double dot(vec a, vec b) { return (a.x*b.x + a.y*b.y); }

def norm_sq(v): return v.x * v.x + v.y * v.y
# double norm_sq(vec v) { return v.x*v.x + v.y*v.y; }

def angle(a, o, b):
  oa = toVec(o, a)
  ob = toVec(o, b)
  return  math.acos(dot(oa, ob) / math.sqrt(norm_sq(oa) * norm_sq(ob)))
# double angle(point a, point o, point b) {  // returns angle aob in rad
#   vec oa = toVec(o, a), ob = toVec(o, b);
#   return acos(dot(oa, ob) / sqrt(norm_sq(oa) * norm_sq(ob))); }

def cross(a, b): return a.x*b.y-a.y*b.x
# double cross(vec a, vec b) { return a.x*b.y - a.y*b.x; }

# // returns the area of polygon P, which is half the cross products
# // of vectors defined by edge endpoints
# double area_alternative(const vector<point> &P) {
#   double ans = 0.0; point O(0.0, 0.0);           // O = the Origin
#   for (int i = 0; i < (int)P.size()-1; ++i)      // sum of signed areas
#     ans += cross(toVec(O, P[i]), toVec(O, P[i+1]));
#   return fabs(ans)/2.0;
# }

def ccw(p, q, r): return (cross(toVec(p,q),toVec(p,r)) > 0)
# note python i used class opperators for tovec (Agis Daniels)
# // note: to accept collinear points, we have to change the `> 0'
# // returns true if point r is on the left side of line pq
# bool ccw(point p, point q, point r) {
#   return cross(toVec(p, q), toVec(p, r)) > 0;
# }

def collinear(p, q, r): return abs(cross(toVec(p, q), toVec(p, r))) < EPS
# // returns true if point r is on the same line as the line pq
# bool collinear(point p, point q, point r) {
#   return fabs(cross(toVec(p, q), toVec(p, r))) < EPS;
# }

def isConvex(P):
  e,s=len(P),1
  if e<4: return False
  t1=ccw(P[0],P[1],P[2]) # first turn
  for i in range(s, e-1):
    if ccw(P[i],P[i+1],P[1 if i+2==n else i+2]) != t1:
      return False
  return True
# // returns true if we always make the same turn
# // while examining all the edges of the polygon one by one
# bool isConvex(const vector<point> &P) {
#   int n = (int)P.size();
#   // a point/sz=2 or a line/sz=3 is not convex  
#   if (n <= 3) return false;
#   bool firstTurn = ccw(P[0], P[1], P[2]);        // remember one result,
#   for (int i = 1; i < n-1; ++i)                 // compare with the others
#     if (ccw(P[i], P[i+1], P[(i+2) == n ? 1 : i+2]) != firstTurn)
#       return false;                              // different -> concave
#   return true;                                   // otherwise -> convex
# }

def insidePolygon(pt, P):
  if len(P)<4: return -1
  n, ans, s=len(P), False, 0.0
  for i in range(n-1):
    a, b=P[i], P[i+1]
    if abs(dist(a, pt) + dist(pt, b) - dist(a, b)) < EPS:
      ans=True
  if ans: return 0
  for i in range(n-1):
    a=angle(P[i], pt, P[i+1])
    s+= a if ccw(pt, P[i], P[i+1]) else -a
  return 1 if abs(s) > math.pi else -1
# // returns 1/0/-1 if point p is inside/on (vertex/edge)/outside of
# // either convex/concave polygon P
# int insidePolygon(point pt, const vector<point> &P) {
#   int n = (int)P.size();
#   if (n <= 3) return -1;                         // avoid point or line
#   bool on_polygon = false;
#   for (int i = 0; i < n-1; ++i)                  // on vertex/edge?
#     if (fabs(dist(P[i], pt) + dist(pt, P[i+1]) - dist(P[i], P[i+1])) < EPS)
#       on_polygon = true;
#   if (on_polygon) return 0;                      // pt is on polygon
#   double sum = 0.0;                              // first = last point
#   for (int i = 0; i < n-1; ++i) {
#     if (ccw(pt, P[i], P[i+1]))
#       sum += angle(P[i], pt, P[i+1]);            // left turn/ccw
#     else
#       sum -= angle(P[i], pt, P[i+1]);            // right turn/cw
#   }
#   return fabs(sum) > M_PI ? 1 : -1;              // 360d->in, 0d->out
# }

def lineInterSectSeg(p,q,A,B):
  a, b, c=B.y-A.y, A.x-B.x, cross(B,A)
  u, v=abs(a*p.x+b*p.y+c), abs(a*q.x+b*q.y+x) 
  return point((p.x*v+q.x*u)/(u+v), (p.y*v+q.y*u)/(u+v))
# // compute the intersection point between line segment p-q and line A-B
# point lineIntersectSeg(point p, point q, point A, point B) {
#   double a = B.y-A.y, b = A.x-B.x, c = B.x*A.y - A.x*B.y;
#   double u = fabs(a*p.x + b*p.y + c);
#   double v = fabs(a*q.x + b*q.y + c);
#   return point((p.x*v + q.x*u) / (u+v), (p.y*v + q.y*u) / (u+v));
# }

def cutPolygon(A, B, Q):
  P=[]
  for i in range(len(Q)):
    l1, l2=cross(toVec(A, B),toVec(A, Q[i])), 0
    if i!=len(Q)-1: l2=cross(toVec(A, B), toVec(A, Q[i+1]))
    if l1>-EPS: P.append(Q[i])
    if l1*l2<-EPS: P.append(lineInterSectSeg(Q[i], Q[i+1], A, B))
  if P and P[-1]!=P[0]: P.append(P[0])
  return P
# // cuts polygon Q along the line formed by point A->point B (order matters)
# // (note: the last point must be the same as the first point)
# vector<point> cutPolygon(point A, point B, const vector<point> &Q) {
#   vector<point> P;
#   for (int i = 0; i < (int)Q.size(); ++i) {
#     double left1 = cross(toVec(A, B), toVec(A, Q[i])), left2 = 0;
#     if (i != (int)Q.size()-1) left2 = cross(toVec(A, B), toVec(A, Q[i+1]));
#     if (left1 > -EPS) P.push_back(Q[i]);         // Q[i] is on the left
#     if (left1*left2 < -EPS)                      // crosses line AB
#       P.push_back(lineIntersectSeg(Q[i], Q[i+1], A, B));
#   }
#   if (!P.empty() && !(P.back() == P.front()))
#     P.push_back(P.front());                      // wrap around
#   return P;
# }

def CH_Graham(Pts):
  P=[p for p in Pts]
  n=len(P)
  if n<4: 
    if P[0]!=P[-1]: P.append(P[0])
    return P
  
  def ccw_cmp(a, b): return ccw(P[0], a, b)
  def cmp_class(f):
    class K:
      def __init__(F, o): F.pt=o
      def __lt__(F, o): return f(F.pt, o.pt)
  P0=P.index(min(P, key=lambda p: (p.y,-p.x)))
  P[0],P[P0]=P[P0],P[0]
  
  P=[P[0]]+sorted(P[1:], key=cmp_class(ccw_cmp))
  
  S, i=[P[-1],P[0]], 2
  while i<n:
    j=len(S)-1
    if not ccw(S[j-1], S[j], P[i]): 
      S.append(P[i]); i+=1
    else: S.pop()
  return S
# vector<point> CH_Graham(vector<point> &Pts) {    // overall O(n log n)
#   vector<point> P(Pts);                          // copy all points
#   int n = (int)P.size();
#   if (n <= 3) {                                  // point/line/triangle
#     if (!(P[0] == P[n-1])) P.push_back(P[0]);    // corner case
#     return P;                                    // the CH is P itself
#   }

#   // first, find P0 = point with lowest Y and if tie: rightmost X
#   int P0 = min_element(P.begin(), P.end())-P.begin();
#   swap(P[0], P[P0]);                             // swap P[P0] with P[0]

#   // second, sort points by angle around P0, O(n log n) for this sort
#   sort(++P.begin(), P.end(), [&](point a, point b) {
#     return ccw(P[0], a, b);                      // use P[0] as the pivot
#   });

#   // third, the ccw tests, although complex, it is just O(n)
#   vector<point> S({P[n-1], P[0], P[1]});         // initial S
#   int i = 2;                                     // then, we check the rest
#   while (i < n) {                                // n > 3, O(n)
#     int j = (int)S.size()-1;
#     if (ccw(S[j-1], S[j], P[i]))                 // CCW turn
#       S.push_back(P[i++]);                       // accept this point
#     else                                         // CW turn
#       S.pop_back();                              // pop until a CCW turn
#   }
#   return S;                                      // return the result
# }

#compressed version of the code below and the link below
#https://en.wikibooks.org/wiki/Algorithm_Implementation/Geometry/Convex_hull/Monotone_chain
def CH_Andrew(ps):
  P, H=sorted(set(ps)), []
  if len(P)<=1: return P # if only one unique point just return point
  def f(B): #f is a mapping of the two loops since its dup code
    for p in P:
      while len(H)>B and not ccw(H[-2], H[-1], p): H.pop()
      H.append(p)
    H.pop()
  f(1); P=P[::-1]; f(len(H)+1); return H #4 line low, rev, up, ret
#c++ implementation below 
# vector<point> CH_Andrew(vector<point> &Pts) {    // overall O(n log n)
#   int n = Pts.size(), k = 0;
#   vector<point> H(2*n);
#   sort(Pts.begin(), Pts.end());                  // sort the points by x/y
#   for (int i = 0; i < n; ++i) {                  // build lower hull
#     while ((k >= 2) && !ccw(H[k-2], H[k-1], Pts[i])) --k;
#     H[k++] = Pts[i];
#   }
#   for (int i = n-2, t = k+1; i >= 0; --i) {       // build upper hull
#     while ((k >= t) && !ccw(H[k-2], H[k-1], Pts[i])) --k;
#     H[k++] = Pts[i];
#   }
#   H.resize(k);
#   return H;
# }

# 6(+1) points, entered in counter clockwise order, 0-based indexing
P = []
P.append(point(1, 1))                            # P0
P.append(point(3, 3))                            # P1
P.append(point(9, 1))                            # P2
P.append(point(12, 4))                           # P3
P.append(point(9, 7))                            # P4
P.append(point(1, 7))                            # P5
P.append(P[0])                                   # loop back, P6 = P0

print("Perimeter = {0:0.2f}".format(perimeter(P))) # 31.64
  # printf("Area = %.2lf\n", area(P));             // 49.00
  # printf("Area = %.2lf\n", area_alternative(P)); // also 49.00
  # printf("Is convex = %d\n", isConvex(P));       // 0 (false)

  # //// the positions of P_out, P_on, P_in with respect to the polygon
  # //7 P5------P_on----P4
  # //6 |                  \
  # //5 |                    \
  # //4 |   P_in              P3
  # //3 |   P1___            /
  # //2 | / P_out \ ___    /
  # //1 P0              P2
  # //0 1 2 3 4 5 6 7 8 9 101112

  # point p_out(3, 2); // outside this (concave) polygon
  # printf("P_out is inside = %d\n", insidePolygon(p_out, P)); // -1
  # printf("P1 is inside = %d\n", insidePolygon(P[1], P)); // 0
  # point p_on(5, 7); // on this (concave) polygon
  # printf("P_on is inside = %d\n", insidePolygon(p_on, P)); // 0
  # point p_in(3, 4); // inside this (concave) polygon
  # printf("P_in is inside = %d\n", insidePolygon(p_in, P)); // 1

  # // cutting the original polygon based on line P[2] -> P[4] (get the left side)
  # //7 P5--------------P4
  # //6 |               |  \
  # //5 |               |    \
  # //4 |               |     P3
  # //3 |   P1___       |    /
  # //2 | /       \ ___ |  /
  # //1 P0              P2
  # //0 1 2 3 4 5 6 7 8 9 101112
  # // new polygon (notice the index are different now):
  # //7 P4--------------P3
  # //6 |               |
  # //5 |               |
  # //4 |               |
  # //3 |   P1___       |
  # //2 | /       \ ___ |
  # //1 P0              P2
  # //0 1 2 3 4 5 6 7 8 9

P = cutPolygon(P[2], P[4], P);
  # printf("Perimeter = %.2lf\n", perimeter(P));   // smaller now, 29.15
  # printf("Area = %.2lf\n", area(P));             // 40.00

  # // running convex hull of the resulting polygon (index changes again)
  # //7 P3--------------P2
  # //6 |               |
  # //5 |               |
  # //4 |   P_out       |
  # //3 |               |
  # //2 |   P_in        |
  # //1 P0--------------P1
  # //0 1 2 3 4 5 6 7 8 9
P = CH_Graham(P);                              #// now this is a rectangle
print(perimeter(P))
  # printf("Perimeter = %.2lf\n", perimeter(P));   // precisely 28.00
  # printf("Area = %.2lf\n", area(P));             // precisely 48.00
  # printf("Is convex = %d\n", isConvex(P));       // true
  # printf("P_out is inside = %d\n", insidePolygon(p_out, P)); // 1
  # printf("P_in is inside = %d\n", insidePolygon(p_in, P)); // 1
